### PR TITLE
Release Molecular Submission Docs

### PR DESCRIPTION
**Pull request by rosibaj**

### CODE REVIEW

File: docs/submission/validating-molecular-data.md [hunk 1]
```diff
@@ -0,0 +1,447 @@
...
+|  **sample_barcode**             | read_groups |                                                           |  According to the SAM specification, this is the expected barcode bases as read by the sequencing machine in the absence of errors.                                                                                                                                                                                                                                          |                      |  
+|  **library_name**               | read_groups |  ![Required](/assets/submission/dictionary-required.svg)  |  Name of a sequencing library made from a molecular sample or a sample pool (multiplex sequencing).                                                                                                                                                                                                                                                                          |                      |
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:293]: Based on schema definition, the following fields are all `Required` to exist in the payload, however some of them accept value of `null`:
```
"required": [
            "library_name",
            "platform_unit",
            "is_paired_end",
            "read_length_r1",
            "file_r1",
            "insert_size",
            "read_length_r2",
            "file_r2",
            "sample_barcode",
            "submitter_read_group_id"
          ]
```

File: docs/submission/validating-molecular-data.md [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+
+|  Payload Field     | Payload Section |  Attribute                                                |  Description                                                                                                                                                                  |  Permissible Values     |  
+|-|-|-|-|-|
+|  **studyId**           |  |  ![Required](/assets/submission/dictionary-required.svg)  |  ARGO `Program ID`, the unique identifier of your program. If you have logged into the ARGO Data Platform, this is the Program ID that you see in the Program Services area.  |                         |  
+|  **analysisType**      |  |  ![Required](/assets/submission/dictionary-required.svg)  |  The type of molecular data that is being submitted in the payload.                                                                                                           |  **sequencing_experiment**  |  
+|  **submitter_sequencing**\_ **experiment_id**  | experiment |  ![Required](/assets/submission/dictionary-required.svg)  |  The unique identifier of the sequencing experiment**.                                                     |                                    |
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:277]: For field `Permissible values`, do we want to add the following according to the schema:
```
String values must meet the regular expression:
^[a-zA-Z0-9]{1}[a-zA-Z0-9\\-_\\.:']{0,98}[a-zA-Z0-9]{1}$
Or `null`
```

File: docs/submission/validating-molecular-data.md [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+|  **analysisType**      |  |  ![Required](/assets/submission/dictionary-required.svg)  |  The type of molecular data that is being submitted in the payload.                                                                                                           |  **sequencing_experiment**  |  
+|  **submitter_sequencing**\_ **experiment_id**  | experiment |  ![Required](/assets/submission/dictionary-required.svg)  |  The unique identifier of the sequencing experiment**.                                                     |                                    |  
+|  **platform**                              | experiment |  ![Required](/assets/submission/dictionary-required.svg)  |  The sequencing platform type used in data generation.                                                   | CAPILLARY, LS454, ILLUMINA, SOLID, HELICOS, IONTORRENT, ONT, PACBIO, Nanopore, BGI |  
+|  **platform_model**                        | experiment |                                                           |  The model number of the sequencing machine used in data generation.                                     |                                    |
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:279]: For `Permissible values`, add
```
- string or `null`
```

File: docs/submission/validating-molecular-data.md [hunk 4]
```diff
@@ -0,0 +1,447 @@
...
+|  **platform_model**                        | experiment |                                                           |  The model number of the sequencing machine used in data generation.                                     |                                    |  
+|  **experimental_strategy**                | experiment |  ![Required](/assets/submission/dictionary-required.svg)  |  The primary experimental method. For sequencing data it refers to how the sequencing library was made.  |  WGS, WXS, RNA-Seq, Bisulfite-Seq  |  
+|  **sequencing_date**                       | experiment |                                                           |  Date sequencing was performed.      |datetime format, for example: 2019-06-16 or  2019-06-16T20:20:39+00:00  |  
+|  **read_group_count**  |  |  ![Required](/assets/submission/dictionary-required.svg)  |  The number of read groups in the molecular files being submitted.                                                                                                               |   |
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:282]: For `Permissible values`, add
- Integer
- Minimum `1`

File: docs/submission/validating-molecular-data.md [hunk 5]
```diff
@@ -0,0 +1,447 @@
...
+|  **sequencing_date**                       | experiment |                                                           |  Date sequencing was performed.      |datetime format, for example: 2019-06-16 or  2019-06-16T20:20:39+00:00  |  
+|  **read_group_count**  |  |  ![Required](/assets/submission/dictionary-required.svg)  |  The number of read groups in the molecular files being submitted.                                                                                                               |   |  
+| **read_group_id_in_bam** | read_groups |   |  Optional field indicating the read group id that is in the BAM.  If submitted, this will be used to map the RD ID in the BAM header to the `submitter_read_group_id` in the payload.  After submission, the RG iD in the payload will be taken as the id for all future headers.  This **cannot** be submitted for FASTQ files. |                                                                      |
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:283]: - String or `null`
- String values must meet the regular expression: `^[a-zA-Z0-9\\-_:\\.']+$`

File: docs/submission/validating-molecular-data.md [hunk 6]
```diff
@@ -0,0 +1,447 @@
...
+| **read_group_id_in_bam** | read_groups |   |  Optional field indicating the read group id that is in the BAM.  If submitted, this will be used to map the RD ID in the BAM header to the `submitter_read_group_id` in the payload.  After submission, the RG iD in the payload will be taken as the id for all future headers.  This **cannot** be submitted for FASTQ files. |                                                                      |  
+|  **submitter_read**\_ **group_id**  | read_groups |  ![Required](/assets/submission/dictionary-required.svg)  |  The unique identifier of a read group; must be unique within each payload.                                                                                                                                                                                                                                                                                         |                      |
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:284]: String values must meet the regular expression: 
`^[a-zA-Z0-9\\-_:\\.]+$`

File: docs/submission/validating-molecular-data.md [hunk 7]
```diff
@@ -0,0 +1,447 @@
...
+|  **submitter_read**\_ **group_id**  | read_groups |  ![Required](/assets/submission/dictionary-required.svg)  |  The unique identifier of a read group; must be unique within each payload.                                                                                                                                                                                                                                                                                         |                      |  
+|  **platform_unit**              | read_groups |  ![Required](/assets/submission/dictionary-required.svg)  |  Unique identifier including the {FLOWCELL_BARCODE}.{LANE}.{SAMPLE_BARCODE}. The {FLOWCELL_BARCODE} refers to the unique identifier for a particular flow cell. The {LANE} indicates the lane of the flow cell and the {SAMPLE_BARCODE} is a sample/library-specific identifier. For non-multiplex sequencing, platform unit and read group have a one-to-one relationship.  |                      |
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:285]: For `Permissible values`, add
- String

File: docs/submission/validating-molecular-data.md [hunk 8]
```diff
@@ -0,0 +1,447 @@
...
+|  **is_paired_end**              | read_groups |  ![Required](/assets/submission/dictionary-required.svg)  |  Indicate if paired-end sequencing was performed.                                                                                                                                                                                                                                                                                                                            |  true, false         |  
+|  **file_r1**                    | read_groups |  ![Required](/assets/submission/dictionary-required.svg)  |  Name of the sequencing file containing reads from the first end of a sequencing run.                                                                                                                                                                                                                                                                                        |                      |
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:287]: For `Permissible values`, add
- String

File: docs/submission/validating-molecular-data.md [hunk 9]
```diff
@@ -0,0 +1,447 @@
...
+|  **file_r1**                    | read_groups |  ![Required](/assets/submission/dictionary-required.svg)  |  Name of the sequencing file containing reads from the first end of a sequencing run.                                                                                                                                                                                                                                                                                        |                      |  
+|  **file_r2**                    | read_groups |  ![Required](/assets/submission/dictionary-required.svg)  |  Name of the sequencing file containing reads from the second end of a paired-end sequencing run. Required if and only if paired-end sequencing was done.                                                                                                                                                                                                                    |                      |
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:288]: For `Permissible values`, add
- String or `null`

File: docs/submission/validating-molecular-data.md [hunk 10]
```diff
@@ -0,0 +1,447 @@
...
+|  **file_r2**                    | read_groups |  ![Required](/assets/submission/dictionary-required.svg)  |  Name of the sequencing file containing reads from the second end of a paired-end sequencing run. Required if and only if paired-end sequencing was done.                                                                                                                                                                                                                    |                      |  
+|  **read_length_r1**            | read_groups |                                                           |  Length of sequencing reads in `file_r1`; this corresponds to the number of sequencing cycles of the first end.                                                                                                                                                                                                                                                              |                      |
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:289]: For Permissible values, add
- Integer or `null`
- Integer minimum value is `20`

File: docs/submission/validating-molecular-data.md [hunk 11]
```diff
@@ -0,0 +1,447 @@
...
+|  **read_length_r1**            | read_groups |                                                           |  Length of sequencing reads in `file_r1`; this corresponds to the number of sequencing cycles of the first end.                                                                                                                                                                                                                                                              |                      |  
+|  **read_length_r2**             | read_groups |                                                           |  Length of sequencing reads in `file_r2`; this corresponds to the number of sequencing cycles of the second end.                                                                                                                                                                                                                                                             |                      |
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:290]: For `Permissible values`, add
- Integer or `null`
- Integer minimum value is `20`

File: docs/submission/validating-molecular-data.md [hunk 12]
```diff
@@ -0,0 +1,447 @@
...
+|  **read_length_r2**             | read_groups |                                                           |  Length of sequencing reads in `file_r2`; this corresponds to the number of sequencing cycles of the second end.                                                                                                                                                                                                                                                             |                      |  
+|  **insert_size**                | read_groups |  ![Required](/assets/submission/dictionary-required.svg)  |  For paired-end sequencing, the average size of sequences between two sequencing ends. Required only for paired-end sequencing.                                                                                                                                                                                                                                              |                      |
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:291]: For `Permissible values`, add
- Integer or `null`
- Integer minimum value is `0`

File: docs/submission/validating-molecular-data.md [hunk 13]
```diff
@@ -0,0 +1,447 @@
...
+|  **insert_size**                | read_groups |  ![Required](/assets/submission/dictionary-required.svg)  |  For paired-end sequencing, the average size of sequences between two sequencing ends. Required only for paired-end sequencing.                                                                                                                                                                                                                                              |                      |  
+|  **sample_barcode**             | read_groups |                                                           |  According to the SAM specification, this is the expected barcode bases as read by the sequencing machine in the absence of errors.                                                                                                                                                                                                                                          |                      |
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:292]: For `Permissible values`, add
- String or `null`

File: docs/submission/validating-molecular-data.md (hunk 1 above)
Comment (lindaxiang) [reply docs/submission/validating-molecular-data.md:293]: For `Permissible values` of `library_name`, add
- String

File: docs/submission/validating-molecular-data.md [hunk 14]
```diff
@@ -0,0 +1,447 @@
...
+
+
+## Validating Metadata Payloads
+
+It is very important that molecular data is submitted with valid metadata. As a helpful tool during metadata preparation, we have prepared a validation client that can be run on your data locally before officially submitting. 
+
+Validation will help you ensure your data is formatted correctly (with accurate identifier assignments between metadata and molecular data files) and that your submission goes smoothly.  It also helps the DCC ensure that the downstream [Analysis Pipelines](/docs/analysis-workflows/analysis-overview) will function seamlessly.  
+
+This validation client will check: 
+- Data submitted in the metadata payload aligns with the BAM/FASTQ files.
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:309]: ---> Metadata in the payload aligns with the BAM/FASTQ files.

File: docs/submission/validating-molecular-data.md [hunk 15]
```diff
@@ -0,0 +1,447 @@
...
+|  **library_name**               | read_groups |  ![Required](/assets/submission/dictionary-required.svg)  |  Name of a sequencing library made from a molecular sample or a sample pool (multiplex sequencing).                                                                                                                                                                                                                                                                          |                      |  
+|  **fileName**       | files |  ![Required](/assets/submission/dictionary-required.svg)  |  Name of the file.                                                                            |                      |
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:294]: For `Permissible values`, add
- String
- String values must meet the regular expression: `^[A-Za-z0-9_\\.\\-\\[\\]\\(\\)]+$`

File: docs/submission/validating-molecular-data.md [hunk 16]
```diff
@@ -0,0 +1,447 @@
...
+|  **library_name**               | read_groups |  ![Required](/assets/submission/dictionary-required.svg)  |  Name of a sequencing library made from a molecular sample or a sample pool (multiplex sequencing).                                                                                                                                                                                                                                                                          |                      |  
+|  **fileName**       | files |  ![Required](/assets/submission/dictionary-required.svg)  |  Name of the file.                                                                            |                      |  
+|  **fileSize**       | files |  ![Required](/assets/submission/dictionary-required.svg)  |  Size of the file, in bytes.                                                                  |                      |
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:295]: For `Permissible values`, add
- Integer
- Integer minimum value is `0`

File: docs/submission/validating-molecular-data.md [hunk 17]
```diff
@@ -0,0 +1,447 @@
...
+|  **library_name**               | read_groups |  ![Required](/assets/submission/dictionary-required.svg)  |  Name of a sequencing library made from a molecular sample or a sample pool (multiplex sequencing).                                                                                                                                                                                                                                                                          |                      |  
+|  **fileName**       | files |  ![Required](/assets/submission/dictionary-required.svg)  |  Name of the file.                                                                            |                      |  
+|  **fileSize**       | files |  ![Required](/assets/submission/dictionary-required.svg)  |  Size of the file, in bytes.                                                                  |                      |  
+|  **fileMd5sum**     | files |  ![Required](/assets/submission/dictionary-required.svg)  |  Compute the md5sum of the file. This must match what is computed when the file is uploaded.  |  |  |
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:296]: For `Permissible values`, add
- String
- String values must meet the regular expression: `^[a-fA-F0-9]{32}$`

**Review by ciaranschutte — APPROVED** [review]

**Comment by rosibaj** [comment]
thanks @lindaxiang i will resolve all of the comments on a new PR in the next 2 days!

File: docs/submission/validating-molecular-data.md [hunk 18]
```diff
@@ -0,0 +1,447 @@
...
+pip3 install . # installs seq-tools
+```
+
+You can verify that the installation worked by checking the version: 
+```bash
+seq-tools -v
+```
+To update the client to the latest version, pull the latest version of the repository and run: 
+```bash
+cd seq-tools
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:354]: we may need to add 
`git pull`

File: docs/submission/validating-molecular-data.md [hunk 19]
```diff
@@ -0,0 +1,447 @@
...
+</Tabs>
+
+<!---  Tabs end here -->
+
+
+### Structured Directory Validation
+One method to quickly complete validation is to structure your data in a pre-determined format that `seq-tools` is expecting. Under a submission directory, compile each submission as a folder, with the JSON payload and the associated data files in the same folder. 
+
+Each submission folder **must** contain:
+- The metadata payload as a `.json` file. Only 1 `.json` file must be present in the submission folder.
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:376]: Maybe change `1` to ` one` is more formal?

File: docs/submission/validating-molecular-data.md [hunk 20]
```diff
@@ -0,0 +1,447 @@
...
+```
+
+</TabItem>
+</Tabs>
+
+<!---  Tabs end here -->
+
+
+### Structured Directory Validation
+One method to quickly complete validation is to structure your data in a pre-determined format that `seq-tools` is expecting. Under a submission directory, compile each submission as a folder, with the JSON payload and the associated data files in the same folder.
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:373]: ---> Under a submission directory(E.g, `molecular-data-submission` in the following example)

File: docs/submission/validating-molecular-data.md [hunk 21]
```diff
@@ -0,0 +1,447 @@
...
+```bash
+seq-tools validate submission-directory/*/*.json  
+```
+
+</TabItem>
+<TabItem value="docker">
+
+```bash
+cd ..
+seq-tools-in-docker validate -d submission-directory/*/*.json
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:405]: Can we just use the folder name in the above example, like:
seq-tools-in-docker validate -d molecular-data-submission/*/*.json

File: docs/submission/validating-molecular-data.md [hunk 22]
```diff
@@ -0,0 +1,447 @@
...
+One method to quickly complete validation is to structure your data in a pre-determined format that `seq-tools` is expecting. Under a submission directory, compile each submission as a folder, with the JSON payload and the associated data files in the same folder. 
+
+Each submission folder **must** contain:
+- The metadata payload as a `.json` file. Only 1 `.json` file must be present in the submission folder.
+- All molecular data files associated to that payload, or symlinks to their location.
+
+For example: 
+![Example Directory Structure](/assets/submission/molecular_submission_validation_directory.png)
+
+Once your data is ready, use `seq-tools` from the submission directory to validate all of the data:
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:382]: Once your data is ready, make sure you are under the parent folder of the submission directory and use `seq-tools` to validate all of the data:

File: docs/submission/validating-molecular-data.md [hunk 23]
```diff
@@ -0,0 +1,447 @@
...
+
+```bash
+seq-tools validate submission-directory/*/*.json  
+```
+
+</TabItem>
+<TabItem value="docker">
+
+```bash
+cd ..
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:404]: `cd ..` is not needed.

File: docs/submission/validating-molecular-data.md [hunk 24]
```diff
@@ -0,0 +1,447 @@
...
+  defaultValue="client"
+  values={[
+    { label: 'Client', value: 'client', },
+    { label: 'Docker', value: 'docker', },
+  ]
+}>
+<TabItem value="client">
+
+```bash
+seq-tools validate submission-directory/*/*.json
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:397]: Can we just use the folder name in the above example, like:
```
seq-tools validate molecular-data-submission/*/*.json
```

File: docs/submission/validating-molecular-data.md [hunk 25]
```diff
@@ -0,0 +1,447 @@
...
+seq-tools-in-docker validate -d submission-directory/*/*.json 
+```
+
+</TabItem>
+</Tabs>
+
+<!---  Tabs end here -->
+
+
+You will see an interactive prompt of the results of validation as the tool runs.  Once completed, view the details of the validation report:
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:414]: ---> Once completed, view the details of the validation reports under the current folder:

File: docs/submission/validating-molecular-data.md [hunk 26]
```diff
@@ -0,0 +1,447 @@
...
+
+```bash
+cat validation_report.PASS.jsonl | jq . | less
+cat validation_report.INVALID.jsonl | jq . | less
+```
+### Scriptable Location Validation
+
+If you do not want to use the structured directory, you can use the '-d' option if data files are located in a different directory than where the metadata payloads are listed.
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:422]: If your data files are located in a different directory than where the metadata payloads are listed, and you do NOT want to re-organize them into the above structured directory, you can use the '-d' option to provide the data files location.

File: docs/submission/validating-molecular-data.md [hunk 27]
```diff
@@ -0,0 +1,447 @@
...
+<!---  Tabs end here -->
+
+
+You will see an interactive prompt of the results of validation as the tool runs.  Once completed, view the details of the validation report:
+
+```bash
+cat validation_report.PASS.jsonl | jq . | less
+cat validation_report.INVALID.jsonl | jq . | less
+```
+### Scriptable Location Validation
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:420]: For this section, do you plan to add an example pic as well?

File: docs/submission/validating-molecular-data.md [hunk 28]
```diff
@@ -0,0 +1,447 @@
...
+If you do not want to use the structured directory, you can use the '-d' option if data files are located in a different directory than where the metadata payloads are listed.
+
+```bash
+seq-tools validate -d ../sequencing-data-directory/ metadata_file_only/sequencing_experiment_payload.json
+```
+Or validate all metadata payloads in a directory at once using a wildcard: 
+```
+seq-tools validate -d ../sequencing-data-directory/ submission-payload-directory/*/*.json  
+seq-tools-in-docker validate -d ../sequencing-data-directory/ submission-payload-directory/*/*.json
+```
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:431]: Do you plan to split into tabs as well? @rosibaj

File: docs/submission/validating-molecular-data.md [hunk 29]
```diff
@@ -0,0 +1,447 @@
...
+
+You can use this method to build your own script to validate payloads in different locations. 
+
+### Understanding the Validation Report
+Validation report summaries can be found in the top level submission directory where you ran `seq-tools`. Reports are separated into three categories:
```
Comment (lindaxiang) [inline docs/submission/validating-molecular-data.md:436]: ---> Validation report summaries can be found in folder where you ran `seq-tools`.

File: docs/submission/validating-molecular-data.md (hunk 28 above)
Comment (rosibaj) [reply docs/submission/validating-molecular-data.md:431]: done

File: docs/submission/validating-molecular-data.md (hunk 1 above)
Comment (rosibaj) [reply docs/submission/validating-molecular-data.md:293]: - all labeled with required now 
- adedd `Any string value` to the `library_name`

File: docs/submission/validating-molecular-data.md (hunk 16 above)
Comment (rosibaj) [reply docs/submission/validating-molecular-data.md:295]: shouldnt this be higher than 0?

**Pull request merged**